### PR TITLE
Eliminate the deprecated PerReadAlleleLikelihoodMap.  Closes #1467.

### LINKED ISSUE #1467: see if PerReadAlleleLikelihoodMap can be / should be refactored out
see https://github.com/broadinstitute/gsa-unstable/issues/831

this came up in review of https://github.com/broadinstitute/gatk/pull/890

### PR DESCRIPTION
**Pull request by davidbenjamin**
@ldgauthier this cleans up something you have previously grumbled about in the annotation code.  Do you feel like reviewing?

### CODE REVIEW

**Comment by ldgauthier** [comment]
Wow, that's a lot of files.  I'll review, but it might take a while.

**Comment by davidbenjamin** [comment]
Thanks!   You'll find the changes to be very repetitive.  It's pretty much just replacing one kind of iteration over most likely / best alleles with another.

**Comment by magicDGS** [comment]
I'm an API user that relies on the `PerReadAlleleLikelihoodMap`, and I would like to know if it is possible to maintain with the `@Deprecated` annotation this class and the utility method for convert it to a `ReadLikelihoods` at least til I get familiar with this class. I'm using a `ReadPileup` to fill up the likelihood map, so if there is some minimal explanation on how to migrate to the non-deprecated class it would be very usuful. Can you help me with this, @davidbenjamin?

**Comment by davidbenjamin** [comment]
@magicDGS Sure!  Can you post some code or point me to a repo?

**Comment by magicDGS** [comment]
Thanks a lot @davidbenjamin! One of the questions that I have is how to add a new read to the `ReadLikelihoods` for a new allele that is found while iterating. A very simple example will be:

```java
    public static PerReadAlleleLikelihoodMap flatPerReadAlleleLikelihoodsFromPileup(final ReadPileup pileup, final Allele refAllele) {
        final PerReadAlleleLikelihoodMap pralm = new PerReadAlleleLikelihoodMap();
        final byte ref = refAllele.getBases()[0];
        for (final PileupElement e : pileup) {
            final Allele current;
            if (e.isDeletion()) {
                current = Allele.SPAN_DEL;
            } else if (e.getBase() == ref) {
                current = refAllele;
            } else {
                current = Allele.create(e.getBase());
            }
            pralm.add(e, current, DEFAULT_FAKE_LIKELIHOOD);
        }
        return pralm;
    }
```

The solution that I found after looking at the class was this one, that it's very complicated:

```java
public static ReadLikelihoods<Allele> flatPerReadAlleleLikelihoodsFromPileup(final ReadPileup pileup, final Allele refAllele, final SAMFileHeader header) {
        final Set<Allele> alleleSet = new TreeSet<Allele>();
        final Map<String, List<GATKRead>> reads = new HashMap<>();
        final byte ref = refAllele.getBases()[0];
        alleleSet.add(refAllele);
        for (final PileupElement e : pileup) {
            if (e.isDeletion()) {
                alleleSet.add(Allele.SPAN_DEL);
            } else if (e.getBase() == ref) {
                alleleSet.add(refAllele);
            } else {
                alleleSet.add(Allele.create(e.getBase()));
            }
            final String sample = ReadUtils.getSampleName(e.getRead(), header);
            List<GATKRead> list = reads.getOrDefault(sample, null);
            if(list == null) {
                list = new ArrayList<>();
                reads.put(sample, list);
            }
            list.add(e.getRead());
        }
        final ReadLikelihoods<Allele> likelihoods = new ReadLikelihoods<>(new IndexedSampleList(reads.keySet()), new IndexedAlleleList<Allele>(alleleSet), reads);
        for(final PileupElement e: pileup) {
            final String sample = ReadUtils.getSampleName(e.getRead(), header);
            final LikelihoodMatrix<Allele> l = likelihoods.sampleMatrix(likelihoods.indexOfSample(sample));
            final int alleleIndex;
            if (e.isDeletion()) {
                alleleIndex = likelihoods.indexOfAllele(Allele.SPAN_DEL);
            } else if (e.getBase() != ref) {
                alleleIndex = likelihoods.indexOfAllele(Allele.create(e.getBase());
            } else {
                alleleIndex = likelihoods.indexOfReference();
            }

            l.set(alleleIndex, l.indexOfRead(e.getRead()), DEFAULT_FAKE_LIKELIHOOD);
        }
        return likelihoods;
    }
```

This example is very simple, but in my case what I need its to assign an unique likelihood to each read after calling the variant for that read. I want to use the variant annotation engine for annotate this likelihood map because it is using this interface.

**Comment by davidbenjamin** [comment]
@magicDGS So, do you just want to construct a `ReadLikelihoods` from a `Pileup` or is there more to it than that?

**Comment by magicDGS** [comment]
Thanks for the quick answer, @davidbenjamin! By the moment is just that, but in the future I would plan to perform more complicating modeling. This is just a proof-of-concept that I'm working on as a starting point.

Could it be possible to implement addition of new alleles, reads for a sample and likelihoods? Something like `add(String sample, GATKRead read, Allele allele, double likelihood)` method, which takes reads/alleles included or not in the initialization?

**Comment by davidbenjamin** [comment]
In `ReadLikelihoods` you can add an `Allele` to all samples and add reads to a single sample.  These methods set a default likelihood.  The current idiom is then to invoke `sampleMatrix()`, which returns the actual matrix, not a copy, and then edit its likelihood values.  This is straightforward but, in my opinion, clumsy.  For example, it would be nice to edit it with a `Visitor` and also to change the `ReadLikelihoods` without invoking `sampleMatrix`.  This is all very doable.

I would not want to add an allele to only one sample because that becomes a different data structure, probably a collection of `LikelihoodMatrix`.  If you need this I think the solution remains to improve the API of existing classes as above.

**Comment by magicDGS** [comment]
I like the first idea the most, not adding it for a concrete sample. The point of the methods that I described is added the likelihood for a single sample, and default likelihood to the others. Probably will be better defined as following: `add(String sample, GATKRead read, Allele allele, double likelihood, double defaultLikelihood)`.

Do you think that you can implement the `Visitor` before removing the `PerReadAlleleLikelihoodMap`? Thanks a lot for all the feedback.

**Comment by davidbenjamin** [comment]
I can do that.  @ldgauthier This won't affect the original commits

**Comment by magicDGS** [comment]
Thanks a lot @davidbenjamin!

**Comment by davidbenjamin** [comment]
@magicDGS After much thought I believe that the code we discussed belongs in your code, not in the GATK.  Here's an outline of some static methods that should meet your needs:
```java
public class Temp {
    @FunctionalInterface
    private interface LikelihoodFunction {
        double apply(final Allele allele, final GATKRead read);
    }

    private static Allele createAllele(final PileupElement e, final byte refBase, final Allele refAllele) {
        return e.isDeletion() ? Allele.SPAN_DEL : (e.getBase() == refBase ? refAllele : Allele.create(e.getBase()));
    }

    public static ReadLikelihoods<Allele> likelihoodsFromPileup(final ReadPileup pileup,
                                                                final Allele refAllele,
                                                                final SAMFileHeader header) {
        final Map<String, List<GATKRead>> readsBySample = pileup.getReads().stream()
                .collect(Collectors.groupingBy(read -> ReadUtils.getSampleName(read, header)));

        final byte refBase = refAllele.getBases()[0];
        final Set<Allele> alleleSet = new TreeSet<>();
        pileup.forEach(e -> alleleSet.add(createAllele(e, refBase, refAllele)));

        return new ReadLikelihoods<>(new IndexedSampleList(readsBySample.keySet()), new IndexedAlleleList<>(alleleSet), readsBySample);

    }

    public static void setLikelihoods(final ReadLikelihoods<Allele> likelihoods, final LikelihoodFunction likelihoodFunction) {
        final int numAlleles = likelihoods.numberOfAlleles();
        for (int sample = 0; sample < likelihoods.numberOfSamples(); sample++) {
            final LikelihoodMatrix<Allele> matrix = likelihoods.sampleMatrix(sample);
            final int numReads = matrix.numberOfReads();
            for (int alleleIndex = 0; alleleIndex < numAlleles; alleleIndex++) {
                final Allele allele = likelihoods.getAllele(alleleIndex);
                for (int readIndex = 0; readIndex < numReads; readIndex++) {
                    final GATKRead read = matrix.getRead(readIndex);
                    matrix.set(alleleIndex, readIndex, likelihoodFunction.apply(allele, read));
                }
            }
        }
    }
}
```

**Comment by davidbenjamin** [comment]
@ldgauthier no new commits; you can go ahead with your review.

**Comment by magicDGS** [comment]
Thanks for the snippet, @davidbenjamin! I will go for something similar, but the main problem that I found in that one is that my toy example is not exactly as you described, because it requires to check the read at the position in the pileup and call the allele again to assign the likelihood only to that one.

I will implement it by myself (it was my plan at the begining), I only asked in my first post if it could be possible to deprecate the `PerReadAlleleLikelihood` instead of completely remove the code to make easier the transition. 

Thank you very much anyway!

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/CoverageUnitTest.java [hunk 1]
```diff
@@ -56,32 +59,52 @@ private VariantContext makeVC() {
 
     @Test
     public void testPerReadAlleleLikelihoodMap(){
-        final PerReadAlleleLikelihoodMap map= new PerReadAlleleLikelihoodMap();
+
+
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/CoverageUnitTest.java]: Nope.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/MappingQualityZeroUnitTest.java [hunk 2]
```diff
@@ -53,47 +50,67 @@ private VariantContext makeVC() {
...
         Assert.assertEquals(annotate.keySet(), Collections.singleton(VCFConstants.MAPPING_QUALITY_ZERO_KEY), "annots");
-        final int n= n1T; //only those are MQ0
-        Assert.assertEquals(annotate.get(VCFConstants.MAPPING_QUALITY_ZERO_KEY), String.valueOf(n));
+        Assert.assertEquals(annotate.get(VCFConstants.MAPPING_QUALITY_ZERO_KEY), String.valueOf(n1T));
 
     }
 
 
     @Test
     public void testPerReadAlleleLikelihoodMapEmpty() throws Exception {
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/MappingQualityZeroUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/QualByDepthUnitTest.java [hunk 3]
```diff
@@ -102,18 +102,30 @@ public void testUsingReads(){
...
             read.setMappingQuality(20);
-            map.add(read, A, -1.0);
-            map.add(read, C, -100.0);  //try to fool it - add another likelihood to same read
-            map.add(read, G, -1000.0);  //and a third one
+            reads.add(read);
+        }
+
+        final Map<String, List<GATKRead>> readsBySample = ImmutableMap.of(sample1, reads);
+        final org.broadinstitute.hellbender.utils.genotyper.SampleList sampleList = new IndexedSampleList(Arrays.asList(sample1));
+        final AlleleList<Allele> alleleList = new IndexedAlleleList<>(Arrays.asList(A, C, G));
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/QualByDepthUnitTest.java]: It was doing nothing.  Got rid of it.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/BaseQualitySumPerAlleleBySampleUnitTest.java [hunk 4]
```diff
@@ -40,16 +39,21 @@ public void testDescriptions() {
...
+
+        final List<GATKRead> reads = new ArrayList<>();
+        final String sample1 = "sample1";
+        final org.broadinstitute.hellbender.utils.genotyper.SampleList sampleList = new IndexedSampleList(Arrays.asList(sample1));
 
         final Allele A = Allele.create("A", true);
         final Allele C = Allele.create("C");
+        final AlleleList<Allele> alleleList = new IndexedAlleleList<>(Arrays.asList(A, C));
 
         final List<Allele> AC = Arrays.asList(A, C);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/BaseQualitySumPerAlleleBySampleUnitTest.java]: The awkwardness is now encapsulated, at least.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/DepthPerAlleleBySampleUnitTest.java [hunk 5]
```diff
@@ -41,41 +39,61 @@ public void testUsingReads(){
...
+            n++;
+        }
+        for (int i = 0; i < readDepthRef; i++) {
+            matrix.set(0, n, -1.0);
+            matrix.set(1, n, -100.0);
+            n++;
+        }
+
+        matrix.set(0, n, -1.0);
+        matrix.set(1, n, -1.1);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/DepthPerAlleleBySampleUnitTest.java]: Big refactoring, no more manually setting likelihoods.

**Comment by davidbenjamin** [comment]
@ldgauthier Here are two commits with everything you requested.

**Comment by davidbenjamin** [comment]
@ldgauthier I *think* you approved without needing it to send back to you, but I'm paranoid and not fully confident with the new github review system.  Also, the changes to the test code to get rid of the duplicated likelihood-setting were pretty big.

**Comment by magicDGS** [comment]
Independently of the result of this, could you wait to merge til #2154 is accepted? Although it is not related to this, it would take me a while to workaround this change in my tool and I require the changes in `ReadPileup` beforehand to check that nothing is broken in my code.

My PR should be accepted soon, so I will appreciate if you can do me this favor. Thanks in advance, @davidbenjamin and  @ldgauthier!

**Comment by davidbenjamin** [comment]
@magicDGS This PR is necessary for my work on Mutect2, but I'm out on Monday and Tuesday anyway.  If #2154 is merged before Wednesday then all is fine; otherwise I can simply re-instate `PerReadAlleleLikelihoodMap` in this PR and delete it in a later PR.

**Comment by ldgauthier** [comment]
I looked at the first new commit and skimmed the second.  I'm assuming
tests still pass.  If you want another set of eyes on the test data
refactor it'll have to wait until Monday.

-L

**Comment by davidbenjamin** [comment]
@ldgauthier I only want an extra set of eyes if it's necessary for approval.  If you want to take my word that the tests are functionally the same but less messy I'll be glad to merge.

**Comment by ldgauthier** [comment]
Definitely less messy and seemingly equivalent at a glance.  👍

**Pull request merged**